File: Android/melody-core/src/main/java/com/melody/core/schema/ComponentStyle.kt
package com.melody.core.schema

import com.charleskorn.kaml.*
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

@Serializable
data class ShadowStyle(
    val x: Double? = null,
    val y: Double? = null,
    val blur: Double? = null,
    val color: String? = null
)

/**
 * Style properties for components.
 * Uses a custom serializer to handle "full" → -1 for size values
 * and kaml YAML decoding.
 */
@Serializable(with = ComponentStyleSerializer::class)
data class ComponentStyle(
    // Value<Double> properties -- accept both literals and {{ expressions }}
    var fontSize: Value<Double>? = null,
    var padding: Value<Double>? = null,
    var paddingTop: Value<Double>? = null,
    var paddingBottom: Value<Double>? = null,
    var paddingLeft: Value<Double>? = null,
    var paddingRight: Value<Double>? = null,
    var paddingHorizontal: Value<Double>? = null,
    var paddingVertical: Value<Double>? = null,
    var margin: Value<Double>? = null,
    var marginTop: Value<Double>? = null,
    var marginBottom: Value<Double>? = null,
    var marginLeft: Value<Double>? = null,
    var marginRight: Value<Double>? = null,
    var marginHorizontal: Value<Double>? = null,
    var marginVertical: Value<Double>? = null,
    var borderRadius: Value<Double>? = null,
    var borderWidth: Value<Double>? = null,
    var width: Value<Double>? = null,
    var height: Value<Double>? = null,
    var minWidth: Value<Double>? = null,
    var minHeight: Value<Double>? = null,
    var maxWidth: Value<Double>? = null,
    var maxHeight: Value<Double>? = null,
    var spacing: Value<Double>? = null,
    var opacity: Value<Double>? = null,
    var cornerRadius: Value<Double>? = null,
    var scale: Value<Double>? = null,
    var rotation: Value<Double>? = null,
    var aspectRatio: Value<Double>? = null,
    var layoutPriority: Value<Double>? = null,

    // Value<String> properties
    var color: Value<String>? = null,
    var backgroundColor: Value<String>? = null,
    var borderColor: Value<String>? = null,

    // Value<ViewAlignment>
    var alignment: Value<ViewAlignment>? = null,

    // Value<Int>
    var lineLimit: Value<Int>? = null,

    // Always-literal properties
    var fontWeight: String? = null,
    var fontDesign: String? = null,
    var shadow: ShadowStyle? = null,
    var animation: String? = null,
    var contentMode: String? = null,
    var overflow: String? = null
)

object ComponentStyleSerializer : KSerializer<ComponentStyle> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("ComponentStyle")

    override fun serialize(encoder: Encoder, value: ComponentStyle) {
        encoder.encodeString("[style]")
    }

    override fun deserialize(decoder: Decoder): ComponentStyle {
        val input = decoder as? YamlInput
            ?: return ComponentStyle()
        val map = input.node as? YamlMap ?: return ComponentStyle()

        val style = ComponentStyle()

        for ((keyNode, valueNode) in map.entries) {
            val key = keyNode.content
            val scalar = (valueNode as? YamlScalar)?.content

            when (key) {
                // Value<Double> numeric properties
                "fontSize" -> style.fontSize = scalar?.let { parseDoubleValue(it) }
                "padding" -> style.padding = scalar?.let { parseDoubleValue(it) }
                "paddingTop" -> style.paddingTop = scalar?.let { parseDoubleValue(it) }
                "paddingBottom" -> style.paddingBottom = scalar?.let { parseDoubleValue(it) }
                "paddingLeft" -> style.paddingLeft = scalar?.let { parseDoubleValue(it) }
                "paddingRight" -> style.paddingRight = scalar?.let { parseDoubleValue(it) }
                "paddingHorizontal" -> style.paddingHorizontal = scalar?.let { parseDoubleValue(it) }
                "paddingVertical" -> style.paddingVertical = scalar?.let { parseDoubleValue(it) }
                "margin" -> style.margin = scalar?.let { parseDoubleValue(it) }
                "marginTop" -> style.marginTop = scalar?.let { parseDoubleValue(it) }
                "marginBottom" -> style.marginBottom = scalar?.let { parseDoubleValue(it) }
                "marginLeft" -> style.marginLeft = scalar?.let { parseDoubleValue(it) }
                "marginRight" -> style.marginRight = scalar?.let { parseDoubleValue(it) }
                "marginHorizontal" -> style.marginHorizontal = scalar?.let { parseDoubleValue(it) }
                "marginVertical" -> style.marginVertical = scalar?.let { parseDoubleValue(it) }
                "borderRadius" -> style.borderRadius = scalar?.let { parseDoubleValue(it) }
                "borderWidth" -> style.borderWidth = scalar?.let { parseDoubleValue(it) }
                "spacing" -> style.spacing = scalar?.let { parseDoubleValue(it) }
                "opacity" -> style.opacity = scalar?.let { parseDoubleValue(it) }
                "cornerRadius" -> style.cornerRadius = scalar?.let { parseDoubleValue(it) }
                "scale" -> style.scale = scalar?.let { parseDoubleValue(it) }
                "rotation" -> style.rotation = scalar?.let { parseDoubleValue(it) }
                "aspectRatio" -> style.aspectRatio = scalar?.let { parseDoubleValue(it) }
                "layoutPriority" -> style.layoutPriority = scalar?.let { parseDoubleValue(it) }

                // Size properties (support "full" -> -1.0)
                "width" -> style.width = scalar?.let { parseSizeValue(it) }
                "height" -> style.height = scalar?.let { parseSizeValue(it) }
                "minWidth" -> style.minWidth = scalar?.let { parseSizeValue(it) }
                "minHeight" -> style.minHeight = scalar?.let { parseSizeValue(it) }
                "maxWidth" -> style.maxWidth = scalar?.let { parseSizeValue(it) }
                "maxHeight" -> style.maxHeight = scalar?.let { parseSizeValue(it) }

                // Value<String> color properties
                "color" -> style.color = scalar?.let { parseStringValue(it) }
                "backgroundColor" -> style.backgroundColor = scalar?.let { parseStringValue(it) }
                "borderColor" -> style.borderColor = scalar?.let { parseStringValue(it) }

                // Value<ViewAlignment>
                "alignment" -> style.alignment = scalar?.let { parseAlignmentValue(it) }

                // Value<Int>
                "lineLimit" -> style.lineLimit = scalar?.let { parseIntValue(it) }

                // Always-literal properties
                "fontWeight" -> style.fontWeight = scalar
                "fontDesign" -> style.fontDesign = scalar
                "animation" -> style.animation = scalar
                "contentMode" -> style.contentMode = scalar
                "overflow" -> style.overflow = scalar

                "shadow" -> {
                    val shadowMap = valueNode as? YamlMap
                    if (shadowMap != null) {
                        style.shadow = ShadowStyle(
                            x = shadowMap.stringAt("x")?.toDoubleOrNull(),
                            y = shadowMap.stringAt("y")?.toDoubleOrNull(),
                            blur = shadowMap.stringAt("blur")?.toDoubleOrNull(),
                            color = shadowMap.stringAt("color")
                        )
                    }
                }
            }
        }

        return style
    }

    /** Parse a numeric value or {{ expression }}. */
    private fun parseDoubleValue(content: String): Value<Double>? {
        Value.extractExpression(content)?.let { return Value.Expression(it) }
        return content.toDoubleOrNull()?.let { Value.Literal(it) }
    }

    /** Parse a size value: number, "full" -> -1.0, or {{ expression }}. */
    private fun parseSizeValue(content: String): Value<Double>? {
        Value.extractExpression(content)?.let { return Value.Expression(it) }
        if (content.lowercase() == "full") return Value.Literal(-1.0)
        return content.toDoubleOrNull()?.let { Value.Literal(it) }
    }

    /** Parse a string value or {{ expression }}. */
    private fun parseStringValue(content: String): Value<String> {
        Value.extractExpression(content)?.let { return Value.Expression(it) }
        return Value.Literal(content)
    }

    /** Parse an alignment value or {{ expression }}. */
    private fun parseAlignmentValue(content: String): Value<ViewAlignment> {
        Value.extractExpression(content)?.let { return Value.Expression(it) }
        return Value.Literal(ViewAlignment.from(content))
    }

    /** Parse an int value or {{ expression }}. */
    private fun parseIntValue(content: String): Value<Int>? {
        Value.extractExpression(content)?.let { return Value.Expression(it) }
        content.toIntOrNull()?.let { return Value.Literal(it) }
        content.toDoubleOrNull()?.toInt()?.let { return Value.Literal(it) }
        return null
    }

    private fun YamlMap.stringAt(key: String): String? {
        return entries.entries.firstOrNull { it.key.content == key }
            ?.value?.let { (it as? YamlScalar)?.content }
    }
}

File: Sources/Core/Schema/Component/ComponentStyle.swift
import Foundation

/// Visual styling properties (layout, spacing, colors, typography) applied to a component.
public struct ComponentStyle: Codable, Sendable, Equatable, Hashable {
    public var fontSize: Value<Double>?
    public var padding: Value<Double>?
    public var paddingTop: Value<Double>?
    public var paddingBottom: Value<Double>?
    public var paddingLeft: Value<Double>?
    public var paddingRight: Value<Double>?
    public var paddingHorizontal: Value<Double>?
    public var paddingVertical: Value<Double>?
    public var margin: Value<Double>?
    public var marginTop: Value<Double>?
    public var marginBottom: Value<Double>?
    public var marginLeft: Value<Double>?
    public var marginRight: Value<Double>?
    public var marginHorizontal: Value<Double>?
    public var marginVertical: Value<Double>?
    public var borderRadius: Value<Double>?
    public var borderWidth: Value<Double>?
    public var width: Value<Double>?
    public var height: Value<Double>?
    public var minWidth: Value<Double>?
    public var minHeight: Value<Double>?
    public var maxWidth: Value<Double>?
    public var maxHeight: Value<Double>?
    public var spacing: Value<Double>?
    public var opacity: Value<Double>?
    public var cornerRadius: Value<Double>?
    public var scale: Value<Double>?
    public var rotation: Value<Double>?
    public var aspectRatio: Value<Double>?
    public var layoutPriority: Value<Double>?

    public var color: Value<String>?
    public var backgroundColor: Value<String>?
    public var borderColor: Value<String>?

    public var alignment: Value<ViewAlignment>?

    public var lineLimit: Value<Int>?

    public var fontWeight: String?
    public var fontDesign: String?
    public var shadow: ShadowStyle?
    public var animation: String?
    public var contentMode: String?
    public var overflow: String?

    private enum CodingKeys: String, CodingKey {
        case fontSize, fontWeight, fontDesign, color, backgroundColor
        case padding, paddingTop, paddingBottom, paddingLeft, paddingRight, paddingHorizontal, paddingVertical
        case margin, marginTop, marginBottom, marginLeft, marginRight, marginHorizontal, marginVertical
        case borderRadius, borderWidth, borderColor
        case width, height, minWidth, minHeight, maxWidth, maxHeight
        case spacing, alignment, shadow, opacity, cornerRadius
        case scale, rotation, animation, aspectRatio, contentMode, overflow, lineLimit, layoutPriority
    }

    public init(
        fontSize: Value<Double>? = nil,
        fontWeight: String? = nil,
        fontDesign: String? = nil,
        color: Value<String>? = nil,
        backgroundColor: Value<String>? = nil,
        padding: Value<Double>? = nil,
        paddingTop: Value<Double>? = nil,
        paddingBottom: Value<Double>? = nil,
        paddingLeft: Value<Double>? = nil,
        paddingRight: Value<Double>? = nil,
        paddingHorizontal: Value<Double>? = nil,
        paddingVertical: Value<Double>? = nil,
        margin: Value<Double>? = nil,
        marginTop: Value<Double>? = nil,
        marginBottom: Value<Double>? = nil,
        marginLeft: Value<Double>? = nil,
        marginRight: Value<Double>? = nil,
        marginHorizontal: Value<Double>? = nil,
        marginVertical: Value<Double>? = nil,
        borderRadius: Value<Double>? = nil,
        borderWidth: Value<Double>? = nil,
        borderColor: Value<String>? = nil,
        width: Value<Double>? = nil,
        height: Value<Double>? = nil,
        minWidth: Value<Double>? = nil,
        minHeight: Value<Double>? = nil,
        maxWidth: Value<Double>? = nil,
        maxHeight: Value<Double>? = nil,
        spacing: Value<Double>? = nil,
        alignment: Value<ViewAlignment>? = nil,
        shadow: ShadowStyle? = nil,
        opacity: Value<Double>? = nil,
        cornerRadius: Value<Double>? = nil,
        scale: Value<Double>? = nil,
        rotation: Value<Double>? = nil,
        animation: String? = nil,
        aspectRatio: Value<Double>? = nil,
        contentMode: String? = nil,
        overflow: String? = nil,
        lineLimit: Value<Int>? = nil,
        layoutPriority: Value<Double>? = nil
    ) {
        self.fontSize = fontSize
        self.fontWeight = fontWeight
        self.fontDesign = fontDesign
        self.color = color
        self.backgroundColor = backgroundColor
        self.padding = padding
        self.paddingTop = paddingTop
        self.paddingBottom = paddingBottom
        self.paddingLeft = paddingLeft
        self.paddingRight = paddingRight
        self.paddingHorizontal = paddingHorizontal
        self.paddingVertical = paddingVertical
        self.margin = margin
        self.marginTop = marginTop
        self.marginBottom = marginBottom
        self.marginLeft = marginLeft
        self.marginRight = marginRight
        self.marginHorizontal = marginHorizontal
        self.marginVertical = marginVertical
        self.borderRadius = borderRadius
        self.borderWidth = borderWidth
        self.borderColor = borderColor
        self.width = width
        self.height = height
        self.minWidth = minWidth
        self.minHeight = minHeight
        self.maxWidth = maxWidth
        self.maxHeight = maxHeight
        self.spacing = spacing
        self.alignment = alignment
        self.shadow = shadow
        self.opacity = opacity
        self.cornerRadius = cornerRadius
        self.scale = scale
        self.rotation = rotation
        self.animation = animation
        self.aspectRatio = aspectRatio
        self.contentMode = contentMode
        self.overflow = overflow
        self.lineLimit = lineLimit
        self.layoutPriority = layoutPriority
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        fontSize = Value.yamlDecode(from: c, key: .fontSize)
        fontWeight = try c.decodeIfPresent(String.self, forKey: .fontWeight)
        fontDesign = try c.decodeIfPresent(String.self, forKey: .fontDesign)
        color = try c.decodeIfPresent(Value<String>.self, forKey: .color)
        backgroundColor = try c.decodeIfPresent(Value<String>.self, forKey: .backgroundColor)
        padding = Value.yamlDecode(from: c, key: .padding)
        paddingTop = Value.yamlDecode(from: c, key: .paddingTop)
        paddingBottom = Value.yamlDecode(from: c, key: .paddingBottom)
        paddingLeft = Value.yamlDecode(from: c, key: .paddingLeft)
        paddingRight = Value.yamlDecode(from: c, key: .paddingRight)
        paddingHorizontal = Value.yamlDecode(from: c, key: .paddingHorizontal)
        paddingVertical = Value.yamlDecode(from: c, key: .paddingVertical)
        margin = Value.yamlDecode(from: c, key: .margin)
        marginTop = Value.yamlDecode(from: c, key: .marginTop)
        marginBottom = Value.yamlDecode(from: c, key: .marginBottom)
        marginLeft = Value.yamlDecode(from: c, key: .marginLeft)
        marginRight = Value.yamlDecode(from: c, key: .marginRight)
        marginHorizontal = Value.yamlDecode(from: c, key: .marginHorizontal)
        marginVertical = Value.yamlDecode(from: c, key: .marginVertical)
        borderRadius = Value.yamlDecode(from: c, key: .borderRadius)
        borderWidth = Value.yamlDecode(from: c, key: .borderWidth)
        borderColor = try c.decodeIfPresent(Value<String>.self, forKey: .borderColor)
        width = Value.yamlDecodeSize(from: c, key: .width)
        height = Value.yamlDecodeSize(from: c, key: .height)
        minWidth = Value.yamlDecodeSize(from: c, key: .minWidth)
        minHeight = Value.yamlDecodeSize(from: c, key: .minHeight)
        maxWidth = Value.yamlDecodeSize(from: c, key: .maxWidth)
        maxHeight = Value.yamlDecodeSize(from: c, key: .maxHeight)
        spacing = Value.yamlDecode(from: c, key: .spacing)
        alignment = try c.decodeIfPresent(Value<ViewAlignment>.self, forKey: .alignment)
        shadow = try c.decodeIfPresent(ShadowStyle.self, forKey: .shadow)
        opacity = Value.yamlDecode(from: c, key: .opacity)
        cornerRadius = Value.yamlDecode(from: c, key: .cornerRadius)
        scale = Value.yamlDecode(from: c, key: .scale)
        rotation = Value.yamlDecode(from: c, key: .rotation)
        animation = try c.decodeIfPresent(String.self, forKey: .animation)
        aspectRatio = Value.yamlDecode(from: c, key: .aspectRatio)
        contentMode = try c.decodeIfPresent(String.self, forKey: .contentMode)
        overflow = try c.decodeIfPresent(String.self, forKey: .overflow)
        lineLimit = Value.yamlDecode(from: c, key: .lineLimit)
        layoutPriority = Value.yamlDecode(from: c, key: .layoutPriority)
    }
}
